### PR TITLE
Extract EnvVariable check invocation from ProcessEvaluationFinishedEventArgs + add Build check events to binglog

### LINKED ISSUE #10431: [Bug] [BuildCheck]: BC0103 used environment variable does not trigger when property is defined in target
### Issue Description

Used environment variable analyzer behavior differs: If assignment happened outside the target the message appears and when in the target it does not. From the [BC0103 rule documentation](https://github.com/dotnet/msbuild/blob/8b03742aafd18a26a5d03cd045e4d24963886503/documentation/specs/BuildCheck/Codes.md#bc0103---used-environment-variable) I am unable to figure out whether this is expected behavior or a bug.

### Steps to Reproduce

1. Create a project `env_var.proj` with following content:

```
<Project>
  <Target Name="Hello">
    <PropertyGroup>
      <ReadFromEnvVariable>$(TEST)</ReadFromEnvVariable>
    </PropertyGroup>
    <Message Importance="High" Text="$(ReadFromEnvVariable)" />
  </Target>
</Project>
```
2. Define the env variable: `$env:TEST="test"`
3. Build with analyzers: `MSBuild.exe .\env_var.proj /m:1 /analyze /t:Hello`

### Expected Behavior

Build outputs 
```
MSBuild.exe .\env_var.proj /m:1 /analyze /t:Hello
MSBuild version 17.12.0-dev-24373-01+8b03742aa for .NET Framework
Build started 7/23/2024 4:43:07 PM.
The BuildCheck is enabled for this build.
C:\Users\alinama\work\shared-ouput\env-var\env_var.proj (3,5): https://aka.ms/buil
dcheck/codes#BC0103 - Property is derived from environment variable: 'TEST' with v
alue: 'test'. Properties should be passed explicitly using the /p option.
Project "C:\Users\alinama\work\shared-ouput\env-var\env_var.proj" on node 1 (Hello
 target(s)).
Hello:
  test
Done Building Project "C:\Users\alinama\work\shared-ouput\env-var\env_var.proj" (H
ello target(s)).


Build succeeded.
    0 Warning(s)
    0 Error(s)
```

### Actual Behavior

Build outputs 
```
Build started 7/23/2024 4:43:26 PM.
The BuildCheck is enabled for this build.
Project "C:\Users\alinama\work\shared-ouput\env-var\env_var.proj" on node 1 (Hello
 target(s)).
Hello:
  test
Done Building Project "C:\Users\alinama\work\shared-ouput\env-var\env_var.proj" (H
ello target(s)).


Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:00.60
```

### Analysis

If assignment happened outside the target, the output consists of the BC0103 message:
```
<Project>
  <PropertyGroup>
    <ReadFromEnvVariable>$(TEST)</ReadFromEnvVariable>
  </PropertyGroup>
  <Target Name="Hello">
    <Message Importance="High" Text="$(ReadFromEnvVariable)" />
  </Target>
</Project>
```
What is interesting is that we have [BC0103 test](https://github.com/dotnet/msbuild/blob/8b03742aafd18a26a5d03cd045e4d24963886503/src/BuildCheck.UnitTests/EndToEndTests.cs#L54) that checks the [assignment in the target](https://github.com/dotnet/msbuild/blob/8b03742aafd18a26a5d03cd045e4d24963886503/src/BuildCheck.UnitTests/TestAssets/SampleAnalyzerIntegrationTest/Project1#L20), and it throws the message BC0103. I am yet unsure what is exactly different in these two cases and brings such difference.

### Versions & Configurations

_No response_

### PR DESCRIPTION
**Pull request by YuliiaKovalova**
Fixes #10431

### Context
When Property is defined inside of the target, it gets evaluated during the [IntrinsicTask](https://github.com/dotnet/msbuild/blob/2f39510e2abb49b27dcb11122088a678c51288cc/src/Build/BackEnd/Components/RequestBuilder/IntrinsicTask.cs#L68) execution, and that happens **after** [ProjectEvaluationFinishedEventArgs](https://github.com/dotnet/msbuild/blob/2f39510e2abb49b27dcb11122088a678c51288cc/src/Build/BuildCheck/Infrastructure/BuildEventsProcessor.cs#L62)

### Changes Made
Extract EnvVariable check invocation from ProcessEvaluationFinishedEventArgs, but make reporting delayed due to dependency on scoping

### Testing
extended existing e2e tests

### CODE REVIEW

**Review by JanKrivanek — COMMENTED** [review]
I might be completely misunderstanding this change - but if we realy on `EnvironmentVariableReadEventArgs` to transfer the info - even for execution. Can we just simply wait for `ProjectFinishedEventArgs` event (similarly as we now wait for `PrjectEvaluationFinishedEventArgs`) to group them.

On a different note - is the groupping of events into single evaluation event correct today? Events from multiple different evaluations can be interleaved - it feels we'd assign them arbitrarily.

**Review by JanKrivanek — COMMENTED** [review]
Few quick comments.

Otherwise looks good - lm know if ready for full review

File: src/Build/BuildCheck/Checks/NoEnvironmentVariablePropertyCheck.cs [hunk 1]
```diff
@@ -39,39 +44,53 @@ public override void Initialize(ConfigurationContext configurationContext)
...
-                }
-                EnvironmentVariableIdentityKey identityKey = new(envVariableData.Key, envVariableData.Value.File, envVariableData.Value.Line, envVariableData.Value.Column);
+                EnvironmentVariableIdentityKey identityKey = new(envVariableData.Key, envVariableData.Value.Location);
                 if (!_environmentVariablesReported.Contains(identityKey))
                 {
-                    if (_isVerboseEnvVarOutput)
+                    string buildCheckResultMessageArgs = _isVerboseEnvVarOutput ? $"'{envVariableData.Key}' with value: '{envVariableData.Value.EnvVarValue}'" : $"'{envVariableData.Key}'";
+
+                    // Scope information is available after evaluation of the project file. If it is not ready, we will report the check later.
+                    if (CheckScopeClassifier.IsScopingReady && CheckScopeClassifier.IsActionInObservedScope(_scope, envVariableData.Value.Location.File, context.Data.ProjectFilePath ?? string.Empty))
```
[the remark this answers is not in the document]
Comment (JanKrivanek) [reply src/Build/BuildCheck/Checks/NoEnvironmentVariablePropertyCheck.cs:79]: Btw. it'd be bit hacky - but would improve perf significantly - how about resorting to stacking only if Scoping is set to `EvaluationCheckScope.WorkTreeImports`?

It is internal builtin check implementation - so we can afford count on internal behavior.

Alternatively the `IsScopingReady` can be a function taking the scope parameter (as for some scopes the classifier is ready immediately)

File: src/Build/BuildCheck/Checks/NoEnvironmentVariablePropertyCheck.cs [hunk 2]
```diff
@@ -39,39 +44,53 @@ public override void Initialize(ConfigurationContext configurationContext)
         foreach (CustomConfigurationData customConfigurationData in configurationContext.CustomConfigurationData)
         {
             bool? isVerboseEnvVarOutput = GetVerboseEnvVarOutputConfig(customConfigurationData, RuleId);
-            _isVerboseEnvVarOutput = isVerboseEnvVarOutput.HasValue && isVerboseEnvVarOutput.Value;           
+            _isVerboseEnvVarOutput = isVerboseEnvVarOutput.HasValue && isVerboseEnvVarOutput.Value;
         }
+
+        CheckScopeClassifier.NotifyOnScopingReadiness += (string? projectFilePath) =>
```
Comment (JanKrivanek) [inline src/Build/BuildCheck/Checks/NoEnvironmentVariablePropertyCheck.cs:50]: How about subscribing a function (not anonymous one) - so that in a handler we can deregister ourselves?
That way we'd avoid double invocations and would be able to get rid of the Disposing logic

File: src/Build/BuildCheck/Infrastructure/BuildEventsProcessor.cs [hunk 3]
```diff
@@ -91,11 +90,18 @@ public TaskKey(BuildEventContext context)
...
         {
-            _evaluatedEnvironmentVariables.Add(envVarName, (envVarValue, file, line, column));
+            _evaluatedEnvironmentVariables.Add(envVarName, (envVarValue, ElementLocation.Create(file, line, column)));
+
+            EnvironmentVariableCheckData checkData =
+               new(file,
+                   checkContext.BuildEventContext?.ProjectInstanceId,
+                   _evaluatedEnvironmentVariables);
+
+            _buildCheckCentralContext.RunEnvironmentVariableActions(checkData, checkContext, ReportResult);
```
Comment (JanKrivanek) [inline src/Build/BuildCheck/Infrastructure/BuildEventsProcessor.cs:104]: I suppose we now want to send the reads one by one - correct?

Then we should get rid of `_evaluatedEnvironmentVariables` alltogether and change the `EnvironmentVariableCheckData` to accomodate single read data

File: src/Build/BuildCheck/Infrastructure/CheckScopeClassifier.cs [hunk 4]
```diff
@@ -7,8 +7,13 @@
 using Microsoft.Build.Shared;
 
 namespace Microsoft.Build.BuildCheck.Infrastructure;
-internal static class CheckScopeClassifier
+
+internal class CheckScopeClassifier : IDisposable
```
Comment (JanKrivanek) [inline src/Build/BuildCheck/Infrastructure/CheckScopeClassifier.cs:11]: Let's not mix static and instance usages.
The dispose on instance clears subscriptions on static class - that can lead to surprises.

I'd suggest to keep this static and having the subscribers being responsible for deregistering (it's a common practice with events anyways)

File: src/Build/BuildCheck/Checks/NoEnvironmentVariablePropertyCheck.cs [hunk 5]
```diff
@@ -39,39 +44,53 @@ public override void Initialize(ConfigurationContext configurationContext)
...
+        {
+            while (_buildCheckResults.Count > 0)
+            {
+                BuildCheckResult result = _buildCheckResults.Pop();
+                if (!CheckScopeClassifier.IsActionInObservedScope(_scope, result.Location.File, projectFilePath ?? string.Empty))
```
Comment (JanKrivanek) [inline src/Build/BuildCheck/Checks/NoEnvironmentVariablePropertyCheck.cs:55]: This feels incorrect. The `projectFile` argument should be indicating in which project file build context did the environment variable read occured - so it needs to be drawn from the environment variable read data. Instead here it is drawn from the project that finished evaluation - which can be completely unrelated.

File: src/Build/BuildCheck/Checks/NoEnvironmentVariablePropertyCheck.cs [hunk 6]
```diff
@@ -39,39 +44,53 @@ public override void Initialize(ConfigurationContext configurationContext)
...
-                            $"'{envVariableData.Key}' with value: '{envVariableData.Value.EnvVarValue}'"));
+                            ElementLocation.Create(envVariableData.Value.Location.File, envVariableData.Value.Location.Line, envVariableData.Value.Location.Column),
+                            buildCheckResultMessageArgs));
                     }
                     else
                     {
-                        context.ReportResult(BuildCheckResult.Create(
+                        _dataContext = context;
+
+                        _buildCheckResults.Push(BuildCheckResult.Create(
```
Comment (JanKrivanek) [inline src/Build/BuildCheck/Checks/NoEnvironmentVariablePropertyCheck.cs:90]: Do we want to create results even for reads that we might be filtering out later on?

File: src/Build/BuildCheck/Checks/NoEnvironmentVariablePropertyCheck.cs (hunk 5 above)
Comment (YuliiaKovalova) [reply src/Build/BuildCheck/Checks/NoEnvironmentVariablePropertyCheck.cs:55]: The trick is that EnvironmentRead event is aware of the file where this variable is detected (it might be an import), the source project is not visible from this event.

File: src/Build/BuildCheck/Checks/NoEnvironmentVariablePropertyCheck.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (YuliiaKovalova) [reply src/Build/BuildCheck/Checks/NoEnvironmentVariablePropertyCheck.cs:79]: Else part pushes everything because we don't know yet if the reported envvariable belongs to the configured  scope :)

[the remark this answers is not in the document]
Comment (YuliiaKovalova) [reply src/Build/BuildCheck/Checks/NoEnvironmentVariablePropertyCheck.cs]: condition for EvaluationCheckScope.WorkTreeImports is added

[the remark this answers is not in the document]
Comment (YuliiaKovalova) [reply src/Build/BuildCheck/Checks/NoEnvironmentVariablePropertyCheck.cs]: now it doesn't look right, need to think more on these conditions.

File: src/Build/BuildCheck/API/BuildCheckResult.cs [hunk 7]
```diff
@@ -32,9 +28,9 @@ public BuildCheckResult(CheckRule checkConfig, IMSBuildElementLocation location,
     internal BuildEventArgs ToEventArgs(CheckResultSeverity severity)
         => severity switch
         {
-            CheckResultSeverity.Suggestion => new BuildCheckResultMessage(this),
-            CheckResultSeverity.Warning => new BuildCheckResultWarning(this, CheckRule.Id),
-            CheckResultSeverity.Error => new BuildCheckResultError(this, CheckRule.Id),
+            CheckResultSeverity.Suggestion => new BuildCheckResultMessage(FormatMessage()),
+            CheckResultSeverity.Warning => new BuildCheckResultWarning(FormatMessage(), CheckRule.Id),
+            CheckResultSeverity.Error => new BuildCheckResultError(FormatMessage(), CheckRule.Id),
```
Comment (JanKrivanek) [inline src/Build/BuildCheck/API/BuildCheckResult.cs:33]: I'd suggest to create a separate PR for this.

What we should really do is to keep passing the `IBuildCheckResult` here, and exposte the info required downstream via that interface and then properly consume it downstream. E.g. the `BuildCheckResultError` ctor doesn't properly populate location information into it's base constructor.

Comment (JanKrivanek) [reply src/Build/BuildCheck/API/BuildCheckResult.cs:33]: I created https://github.com/dotnet/msbuild/issues/10529

File: src/Build/BuildCheck/Checks/NoEnvironmentVariablePropertyCheck.cs [hunk 8]
```diff
@@ -39,44 +43,44 @@ public override void Initialize(ConfigurationContext configurationContext)
...
         }
+
+        CheckScopeClassifier.NotifyOnScopingReadiness += HandleScopeReadiness;
     }
 
-    public override void RegisterActions(IBuildCheckRegistrationContext registrationContext) => registrationContext.RegisterEvaluatedPropertiesAction(ProcessEnvironmentVariableReadAction);
+    public override void RegisterActions(IBuildCheckRegistrationContext registrationContext) => registrationContext.RegisterEnvironmentVariableReadAction(ProcessEnvironmentVariableReadAction);
 
-    private void ProcessEnvironmentVariableReadAction(BuildCheckDataContext<EvaluatedPropertiesCheckData> context)
+    public override void Dispose() => CheckScopeClassifier.NotifyOnScopingReadiness -= HandleScopeReadiness;
```
Comment (JanKrivanek) [inline src/Build/BuildCheck/Checks/NoEnvironmentVariablePropertyCheck.cs:54]: Do we expect and need more then one invocation of the event? If not, then we can deregister right away in the handler

File: src/Build/BuildCheck/Checks/NoEnvironmentVariablePropertyCheck.cs (hunk 6 above)
Comment (JanKrivanek) [reply src/Build/BuildCheck/Checks/NoEnvironmentVariablePropertyCheck.cs]: I still see creation of the results that might be filtered later on

File: src/Build/BuildCheck/Checks/NoEnvironmentVariablePropertyCheck.cs [hunk 9]
```diff
@@ -39,44 +43,44 @@ public override void Initialize(ConfigurationContext configurationContext)
...
-                    {
-                        context.ReportResult(BuildCheckResult.Create(
-                            SupportedRule,
-                            ElementLocation.Create(envVariableData.Value.File, envVariableData.Value.Line, envVariableData.Value.Column),
-                            $"'{envVariableData.Key}'"));
-                    }
-
-                    _environmentVariablesReported.Add(identityKey);
-                }
+                _dataContext ??= context;
```
Comment (JanKrivanek) [inline src/Build/BuildCheck/Checks/NoEnvironmentVariablePropertyCheck.cs:66]: The context is valid for particular invocation. In case of multiple projects with multiple configuration the context will differ. Using mismatched context can lead to strange bugs (e.g. results having wrong severity for a specific project).

tl;dr; - we'll need to queue the context together with the data

File: src/Build/BuildCheck/Infrastructure/BuildCheckManagerProvider.cs [hunk 10]
```diff
@@ -339,18 +339,25 @@ public void ProcessEvaluationFinishedEventArgs(
 
             _buildEventsProcessor
                 .ProcessEvaluationFinishedEventArgs(checkContext, evaluationFinishedEventArgs, propertiesLookup);
+
+            CheckScopeClassifier.RaiseNotifyOnScopingReadiness();
```
Comment (JanKrivanek) [inline src/Build/BuildCheck/Infrastructure/BuildCheckManagerProvider.cs:343]: For the in-proc node and worker nodes this happens after first evaluation is done:

https://github.com/dotnet/msbuild/blob/f422d8d7dfe0a7115b11b31470215ad6b7723138/src/Build/BackEnd/Components/RequestBuilder/RequestBuilder.cs#L1168

How about adding the state+event to `FileClassifier` - there flipping the state on first `RegisterKnownImmutableLocations` call, and in the `CheckScopeClassifier`  just passing through the chack and event subscription to the `FileClassifier`?

Something like (pseudocode):
```csharp
internal class FileClassifier
{
   private bool _isInitialized;
   public bool IsInitialized {get => _isInitialized; private set { if(!_isInitialized){ /* invoke event */ } _isInitialized = value; } }
   public event Action? OnInitialized;

   public void RegisterKnownImmutableLocations(/*...*/)
  {
    // ...
   this.IsInitialized = true;
  }
}

internal static class CheckScopeClassifier
{
   internal bool IsInitialized => FileClassifier.Shared.IsInitialized;
   internal static event Action? OnInitialized
  {
    add => FileClassifier.Shared.OnInitialized += value;
    remove => FileClassifier.Shared.OnInitialized -= value;
  }
}
```

File: src/Build/BuildCheck/Infrastructure/BuildCheckManagerProvider.cs [hunk 11]
```diff
@@ -339,18 +339,25 @@ public void ProcessEvaluationFinishedEventArgs(
...
             {
-                _buildEventsProcessor.ProcessEnvironmentVariableReadEventArgs(
-                    evr.EnvironmentVariableName,
-                    evr.Message ?? string.Empty,
-                    evr.File,
-                    evr.LineNumber,
-                    evr.ColumnNumber);
+                TryGetProjectFullPath(checkContext.BuildEventContext, out string projectPath);
+
+                if (!string.IsNullOrEmpty(projectPath))
```
Comment (JanKrivanek) [inline src/Build/BuildCheck/Infrastructure/BuildCheckManagerProvider.cs:352]:
```suggestion
                if (!TryGetProjectFullPath(checkContext.BuildEventContext, out string projectPath))
```

File: src/Build/BuildCheck/OM/EnvironmentVariableCheckData.cs [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    internal EnvironmentVariableCheckData(
+        string projectFilePath,
+        int? projectConfigurationId,
+        (string EnvVarKey, string EnvVarValue, IMSBuildElementLocation Location) evaluatedEnvironmentVariable)
+        : base(projectFilePath, projectConfigurationId) => EvaluatedEnvironmentVariable = evaluatedEnvironmentVariable;
+
+    /// <summary>
+    /// Gets the evaluated environment variable.
+    /// </summary>
+    public (string EnvVarKey, string EnvVarValue, IMSBuildElementLocation Location) EvaluatedEnvironmentVariable { get; }
```
Comment (JanKrivanek) [inline src/Build/BuildCheck/OM/EnvironmentVariableCheckData.cs:22]: Just curious - why is this a tupple and not separate properties?

**Review by JanKrivanek — CHANGES_REQUESTED** [review]
Looks very good!

I have some pesky comments :-) - mainly let's make sure there is no mismatch when reporting the results, plus lets minimise the occations where we need to pre-buffer the data

File: src/Build/BuildCheck/OM/EnvironmentVariableCheckData.cs (hunk 12 above)
Comment (YuliiaKovalova) [reply src/Build/BuildCheck/OM/EnvironmentVariableCheckData.cs:22]: it was handy for the list of envvars before, but now I can change it

File: src/Framework/FileClassifier.cs [hunk 13]
```diff
@@ -36,6 +36,30 @@ namespace Microsoft.Build.Framework
...
+
+        /// <summary>
+        ///  Tracks whether the immutable paths have been initialized.
+        /// </summary>
+        public bool IsImmutablePathsInitialized
+        {
+            get => _isImmutablePathsInitialized;
+            private set
+            {
+                if (!_isImmutablePathsInitialized)
```
Comment (JanKrivanek) [inline src/Framework/FileClassifier.cs:54]: nit: the codepath allows setting to `false` as well. So just to be defensive

```suggestion
                if (!_isImmutablePathsInitialized && value)
```

File: src/Build/BuildCheck/Checks/NoEnvironmentVariablePropertyCheck.cs [hunk 14]
```diff
@@ -39,44 +41,41 @@ public override void Initialize(ConfigurationContext configurationContext)
...
+
+            // Scope information is available after evaluation of the project file. If it is not ready, we will report the check later.
+            if (!CheckScopeClassifier.IsScopingReady(_scope))
+            {
+                _buildCheckResults.Enqueue(
+                    (context.Data.ProjectFilePath,
+                    BuildCheckResult.Create(
+                        SupportedRule,
+                        context.Data.EnvironmentVariableLocation,
+                        buildCheckResultMessageArgs),
```
Comment (JanKrivanek) [inline src/Build/BuildCheck/Checks/NoEnvironmentVariablePropertyCheck.cs:67]: nit: The `SupportedRule` and `buildCheckResultMessageArgs` can be derived from the instance context - so the only data we really need to store is the location (instead of the whole result) - and we would safe couple dozens bytes in memory per each call here.

**Review by JanKrivanek — APPROVED** [review]
This looks great and very solid!

Thank you for all the iterations!

File: src/Build/Logging/BinaryLogger/BinaryLogger.cs [hunk 15]
```diff
@@ -76,8 +76,9 @@ public sealed class BinaryLogger : ILogger
         // version 22:
         //    - extend EnvironmentVariableRead with location where environment variable was used.
         // version 23:
-        //    - new record kind: BuildSubmissionStartedEventArgs
-        // This should be never changed.
```
Comment (maridematte) [inline src/Build/Logging/BinaryLogger/BinaryLogger.cs:80]: Comment about `ForwardCompatibilityMinimalVersion` accidentally removed.

File: src/BuildCheck.UnitTests/EndToEndTests.cs [hunk 16]
```diff
@@ -236,9 +235,12 @@ public void NoEnvironmentVariableProperty_Test(bool? customConfigEnabled, string
             customConfigData);
 
         string output = RunnerUtilities.ExecBootstrapedMSBuild(
-            $"{Path.GetFileName(projectFile.Path)} /m:1 -nr:False -restore -check", out bool success, false, _env.Output);
+            $"{Path.GetFileName(projectFile.Path)} /m:1 -nr:False -restore -check", out bool success, false, _env.Output, timeoutMilliseconds: 12000_000);
```
Comment (maridematte) [inline src/BuildCheck.UnitTests/EndToEndTests.cs:238]: From one test the timeout was removed and in this one it was added, which one are we going with? Yes timeout, no timeout?

File: src/Build/BuildCheck/Infrastructure/CheckScopeClassifier.cs [hunk 17]
```diff
@@ -7,8 +7,26 @@
...
+
 internal static class CheckScopeClassifier
 {
+    static CheckScopeClassifier() =>
+        FileClassifier.Shared.OnImmutablePathsInitialized += () =>
+        {
+            NotifyOnScopingReadiness?.Invoke();
+
+            FileClassifier.Shared.OnImmutablePathsInitialized -= () => NotifyOnScopingReadiness?.Invoke();
+        };
```
Comment (maridematte) [inline src/Build/BuildCheck/Infrastructure/CheckScopeClassifier.cs:19]: I think that this is very hard to read right now. Could you please simplify the coding style?

**Review by maridematte — COMMENTED** [review]
Looks good to me, just some nits around code styling :)

File: src/BuildCheck.UnitTests/EndToEndTests.cs (hunk 16 above)
Comment (JanKrivanek) [reply src/BuildCheck.UnitTests/EndToEndTests.cs]: No timeout is preferable (the utils will fallback to default 30 seconds timeout).
Historicaly some of the unittests were adding longer timeouts as we had a perf issue and it was an attempt to workaround - this is not legitimate reason anymore.
Sometimes (this is the case here likely) - devs are adding high timeouts to be able to debug bootstrapped msbuild, without runner utils killing it after timeout. The more prefered way is setting the `MSBuildDebugUnitTests` env var - which then stops enforcing any timeout at all. The env var can be set as a global variable on dev machine - so one doesnt have to care about it anymore

**Review by maridematte — APPROVED** [review]
